src/sgame/botlib/bot_load.cpp

```diff
@@ -79,101 +98,175 @@ void BotSaveOffMeshConnections( NavData_t *nav )
+			(int) nav->process.con.verts[ ( 6 * n ) + 0 ],
+			(int) nav->process.con.verts[ ( 6 * n ) + 1 ],
+			(int) nav->process.con.verts[ ( 6 * n ) + 2 ],
+			(int) nav->process.con.verts[ ( 6 * n ) + 3 ],
+			(int) nav->process.con.verts[ ( 6 * n ) + 4 ],
+			(int) nav->process.con.verts[ ( 6 * n ) + 5 ],
+			(int) nav->process.con.rad[ n ],
+			nav->process.con.flags[ n ],
+			nav->process.con.areas[ n ],
+			nav->process.con.dirs[ n ] );
+
+		trap_FS_Write( s, strlen( s ), f );
+	}
 
 	trap_FS_FCloseFile( f );
+
+	BotPrintOffMeshConnections( nav );
 }
 
 static void BotLoadOffMeshConnections( const char *filename, NavData_t *nav )
 {
+	nav->process.con.offMeshConCount = 0;
+
 	char filePath[ MAX_QPATH ];
 	fileHandle_t f = 0;
 
 	std::string mapname = Cvar::GetValue("mapname");
 	Com_sprintf( filePath, sizeof( filePath ), "maps/%s-%s.navcon", mapname.c_str(), filename );
-	G_FOpenGameOrPakPath( filePath, f );
+	int len = G_FOpenGameOrPakPath( filePath, f );
 
 	if ( !f )
 	{
 		return;
 	}
 
-	OffMeshConnectionHeader header;
-	trap_FS_Read( &header, sizeof( header ), f );
+	int version;
+	trap_FS_Read( &version, sizeof( version ), f );
+	len -= sizeof( version );
 
-	header.version = LittleLong( header.version );
-	header.numConnections = LittleLong( header.numConnections );
+	version = LittleLong( version );
 
-	if ( header.version != NAVMESHCON_VERSION )
+	// Old binary format.
+	if ( version == 2 )
 	{
-		trap_FS_FCloseFile( f );
-		return;
-	}
+		Log::Debug("Loading format version %d navcon file %s", version, filePath );
 
-	int conCount = header.numConnections;
+		int conCount;
+		trap_FS_Read( &conCount, sizeof( conCount ), f );
 
-	if ( conCount > nav->process.con.MAX_CON )
-	{
+		conCount = LittleLong( conCount );
+
+		if ( conCount > nav->process.con.MAX_CON )
+		{
+			Log::Warn( "Too much connections in navcon file %s", filePath );
+			trap_FS_FCloseFile( f );
+			return;
+		}
+
+		nav->process.con.offMeshConCount = conCount;
+
+		trap_FS_Read( nav->process.con.verts, sizeof( float ) * 6 * conCount, f );
+		SwapArray( nav->process.con.verts, conCount * 6 );
+
+		trap_FS_Read( nav->process.con.rad, sizeof( float ) * conCount, f );
+		SwapArray( nav->process.con.rad, conCount );
+
+		trap_FS_Read( nav->process.con.flags, sizeof( unsigned short ) * conCount, f );
+		SwapArray( nav->process.con.flags, conCount );
+
+		trap_FS_Read( nav->process.con.areas, sizeof( unsigned char ) * conCount, f );
+		trap_FS_Read( nav->process.con.dirs, sizeof( unsigned char ) * conCount, f );
+
+		// userids is loaded but unused.
+		trap_FS_Read( nav->process.con.userids, sizeof( unsigned int ) * conCount, f );
+		SwapArray( nav->process.con.userids, conCount );
+
+		Log::Debug( "Loaded %d connections from navcon file %s", conCount, filePath );
 		trap_FS_FCloseFile( f );
+
+		BotPrintOffMeshConnections( nav );
 		return;
 	}
+	// New text format, check "navc" from "navcon".
+	else if ( version == 1668702574 )
```

Comment:
For a text format we shouldn't require that the first token is at the exact beginning of the file since whitespace is not supposed to be significant. Just start treating it as text with `sscanf` etc.

Reply:
how can I restart a `read` from the start of the file without closing/reopening the file?

Reply:
Also I don't think “white space are not supposed to be significant” is a good idea there, it's not really supposed to be written by hand like a material or a map metadata file, and handling whitespace agnosticism just adds extra complexity that is probably overkill and useless, we would also have to avoid `sscanf`… This being a text format is primarily made to be diffable/mergeable, I don't mind if the format is strict.

Reply:
We have `trap_FS_Seek`